src/category_theory/shift.lean

```diff
@@ -1,46 +1,346 @@
+  (shift_add X i j).hom ≫ eq_to_hom (by rw h) = eq_to_hom (by rw h) ≫ (shift_add X i' j).hom :=
+by { cases h, rw [eq_to_hom_refl, eq_to_hom_refl, category.id_comp, category.comp_id] }
+
+@[reassoc] lemma shift_add_hom_comp_eq_to_hom₂ (i j j' : A) (h : j = j') :
+  (shift_add X i j).hom ≫ eq_to_hom (by rw h) = eq_to_hom (by rw h) ≫ (shift_add X i j').hom :=
+by { cases h, rw [eq_to_hom_refl, eq_to_hom_refl, category.id_comp, category.comp_id] }
+
+@[reassoc] lemma shift_add_hom_comp_eq_to_hom₁₂ (i j i' j' : A) (h₁ : i = i') (h₂ : j = j') :
+  (shift_add X i j).hom ≫ eq_to_hom (by rw [h₁, h₂]) =
+    eq_to_hom (by rw [h₁, h₂]) ≫ (shift_add X i' j').hom :=
+by { cases h₁, cases h₂, rw [eq_to_hom_refl, eq_to_hom_refl, category.id_comp, category.comp_id] }
+
+@[reassoc] lemma eq_to_hom_comp_shift_add_inv₁ (i i' j : A) (h : i = i') :
+  eq_to_hom (by rw h) ≫ (shift_add X i' j).inv = (shift_add X i j).inv ≫ eq_to_hom (by rw h) :=
+by rw [iso.comp_inv_eq, category.assoc, iso.eq_inv_comp, shift_add_hom_comp_eq_to_hom₁]
+
+@[reassoc] lemma eq_to_hom_comp_shift_add_inv₂ (i j j' : A) (h : j = j') :
+  eq_to_hom (by rw h) ≫ (shift_add X i j').inv = (shift_add X i j).inv ≫ eq_to_hom (by rw h) :=
+by rw [iso.comp_inv_eq, category.assoc, iso.eq_inv_comp, shift_add_hom_comp_eq_to_hom₂]
+
+@[reassoc] lemma eq_to_hom_comp_shift_add_inv₁₂ (i j i' j' : A) (h₁ : i = i') (h₂ : j = j') :
+  eq_to_hom (by rw [h₁, h₂]) ≫ (shift_add X i' j').inv =
+    (shift_add X i j).inv ≫ eq_to_hom (by rw [h₁, h₂]) :=
+by rw [iso.comp_inv_eq, category.assoc, iso.eq_inv_comp, shift_add_hom_comp_eq_to_hom₁₂]
+
+lemma shift_shift' (i j : A) :
+  f⟦i⟧'⟦j⟧' = (shift_add X i j).inv ≫ f⟦i + j⟧' ≫ (shift_add Y i j).hom :=
+by { symmetry, apply nat_iso.naturality_1 }
+
+variables (A)
+
+/-- Shifting by zero is the identity functor. -/
+abbreviation shift_zero  :
+  X⟦0⟧ ≅ X := (shift_functor_zero C A).app _
+
+lemma shift_zero' :
+  f⟦(0 : A)⟧' = (shift_zero A X).hom ≫ f ≫ (shift_zero A Y).inv :=
+by { symmetry, apply nat_iso.naturality_2 }
+
+end add_monoid
+
+section opaque_eq_to_iso
+
+variables {ι : Type*} {i j k : ι}
 
-example {X Y : C} (f : X ⟶ Y) : X⟦1⟧ ⟶ Y⟦1⟧ := f⟦1⟧'
-example {X Y : C} (f : X ⟶ Y) : X⟦-2⟧ ⟶ Y⟦-2⟧ := f⟦-2⟧'
+/-- This definition is used instead of `eq_to_iso` so that the proof of `i = j` is visible
+to the simplifier -/
+def opaque_eq_to_iso (h : i = j) : @iso (discrete ι) _ i j := eq_to_iso h
+
+@[simp]
+lemma opaque_eq_to_iso_symm (h : i = j) :
+  (opaque_eq_to_iso h).symm = opaque_eq_to_iso h.symm := rfl
+
+@[simp]
+lemma opaque_eq_to_iso_inv (h : i = j) :
+  (opaque_eq_to_iso h).inv = (opaque_eq_to_iso h.symm).hom := rfl
+
+@[simp, reassoc]
+lemma map_opaque_eq_to_iso_comp_app (F : discrete ι ⥤ C ⥤ C) (h : i = j) (h' : j = k) (X : C) :
+  (F.map (opaque_eq_to_iso h).hom).app X ≫ (F.map (opaque_eq_to_iso h').hom).app X =
+    (F.map (opaque_eq_to_iso $ h.trans h').hom).app X := by { delta opaque_eq_to_iso, simp }
```

Comment:
Please move these to the appropriate file (presumably around the definition of `discrete`).

Reply:
I think this and the simp lemma below is quite specific to this particular use case.
Or do we want a general variant of `eq_to_hom` where the simp lemmas aren't that aggressive?

Reply:
I think it's okay to leave these here for now, but if we end up needing a similar trick somewhere else later on, we should indeed try to develop a general variant of `eq_to_hom` with adjusted simp lemmas, as you described.